Add events to status manager and controller

alonSadan opened the pull request:
<!-- Thanks for sending a pull request!

Before you click the 'Create pull request' make sure that:
- This PR introduces a single feature of fix, just one
- This PR does not leave the master branch broken
- Every commit in this PR has a commit message explaining what do you change,
  why and what is the outcome
- If your change introduces a complex concept or a change to user interaction
  with the project or the application, make sure to document it

If you don't comply with these rules, you waste your energy, time of reviewers
and cause suffering of future generations.
-->

**What this PR does / why we need it**:

In order to  bring CNAO to meet operators conventional best practices,
It should use k8s events to describe actions that are being taken.

In this PR:

- A new event emitter is introduced, to easily produce events.
the emitter has to be initiated and then can be used to create events.   
Right now it only support associating the event to a runtime object,
but can be improved in the future to also support associating events to
CSVs and the self pod.

- Add events to log the deployment of all (not each) component.
Three types of events will be logged:  Available, Progressing and Failed.
The events will be associated with the networkAddonsConfig object. and 
will be produced in statusManager.

- Add an event on changes made to networkAddonsConfig.Spec,
to log any addition or removal of a network component.

- Added some info about the newly added events, and how to get them to the readme.

- Removed redundant error formatting from Wrapf() functions.  
   

events output( `kubectl describe kubectl describe networkaddonsconfig/cluste`) : 

Modified: 
```
  Type    Reason    Age   From                             Message
  ----    ------    ----  ----                             -------
  Normal  Modified  92s   cluster-network-addons-operator  Config spec was modified
```

Progressing: 
```
  Type    Reason    Age   From                             Message
  ----    ------    ----  ----                             -------
Normal  Progressing  86s (x4 over 88s)  cluster-network-addons-operator  Operator is progressing

```

Available: 
```
  Type    Reason    Age   From                             Message
  ----    ------    ----  ----                             -------
 Normal  Available    2s (x7 over 63s)   cluster-network-addons-operator  All applied componenets are available
```

Failed( this is an example, could be many Failing event types. in this example I gave illegal range for KMP ):
```
 Type    Reason    Age   From                             Message
  ----    ------    ----  ----                             -------
Warning  InvalidConfiguration  91s (x17 over 7m18s)  cluster-network-addons-operator  Operator failed: Human interaction is 
needed, please fix the desired configuration
```

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If no release note is required, just write "NONE".
-->

```release-note
k8s events were added to log failed, progressing, successful and modified deployment  
```

alonSadan commented:
/assign @RamLavi

qinqon commented on pkg/names/names.go:
```diff
@@ -12,3 +12,6 @@ const APPLIED_PREFIX = "cluster-networks-addons-operator-applied-"
 const REJECT_OWNER_ANNOTATION = "networkaddonsoperator.network.kubevirt.io/rejectOwner"
+
+// ClusterNetworkOperatorName is the name of the cluster-network-addons-operator resource that will be reconciled
+const ClusterNetworkAddonsOperatorName = "cluster-network-addons-operator"
```
Looks like const are UPPER_CASE at CNAO

qinqon commented on pkg/util/k8s/event_emmiter.go:
```diff
@@ -0,0 +1,37 @@
+package k8s
```
Move this to `pkg/eventemitter/eventemitter.go` so we do `eventemitter.New()`

Also are we really sure that we need the interface ? I suspect is for unit testing if this is the case maybe we can mock `GetEventRecorderFor` somehow passing a fake manager.

qinqon replied:
Nah EventEmitter feels right.

Well maybe the most important part is tu split where do we trigger the events and how do we format them, for that I sould say that we create some high level functions at `EventEmitter` that reads the networkaddons config and format the events as needed

```golang
func EmmitAvailable(configName string)
func EmmitProgressing(configName string)
func EmmitFailing(configName string)
func EmmitWarning(configName string, reason string, message string)
```

alonSadan replied:
> Let's create some high level functions with event formating so we split responsabilities, added a comment to it.

Thanks @qinqon 
I added the external function. I hope that's what you meant.

I didn't understand what is EmitWarning. can you elaborate?

I agree we need to think together about were we wnat to place the events. this is my suggestion, but I am not really sure of it...

qinqon reviewed (changes requested):
First pass

phoracek commented on pkg/controller/statusmanager/status_manager.go:
```diff
@@ -214,6 +233,12 @@ func (status *StatusManager) SetNotFailing(level StatusLevel) {
+	if err := status.client.Get(context.TODO(), types.NamespacedName{Name: names.OPERATOR_CONFIG}, config); err != nil {
+		return
+	}
+	eventEmitter.EmitEvent(config, corev1.EventTypeNormal, "Progressing", fmt.Sprintf("Operator is progressing; Now on level %d", level))
```
The level is internal detail, it would not provide user any extra information

RamLavi commented on pkg/eventemitter/eventemitter.go:
```diff
@@ -0,0 +1,33 @@
+package eventemitter
+
```
commit typo `to  bring` (double space)

alonSadan replied:
Done.

RamLavi commented on pkg/eventemitter/eventemitter.go:
```diff
@@ -0,0 +1,33 @@
+package eventemitter
+
+import (
+	"github.com/kubevirt/cluster-network-addons-operator/pkg/names"
```
please put self imports in a separate section:
```
import (
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/client-go/tools/record"
	"sigs.k8s.io/controller-runtime/pkg/manager"

	"github.com/kubevirt/cluster-network-addons-operator/pkg/names"
```

alonSadan replied:
Done

RamLavi commented on pkg/controller/statusmanager/status_manager.go:
```diff
@@ -346,8 +366,8 @@ func (status *StatusManager) SetFromPods() {
-	config := &opv1alpha1.NetworkAddonsConfig{}
-	if err := status.client.Get(context.TODO(), types.NamespacedName{Name: names.OPERATOR_CONFIG}, config); err != nil {
+	config := &opv1alpha1.NetworkAddonsConfig{ObjectMeta: metav1.ObjectMeta{Name: status.name}}
+	if err := status.client.Get(context.TODO(), types.NamespacedName{Name: status.name}, config); err != nil {
```
why is this change needed?

alonSadan replied:
I don't think there is much difference, I just wanted it be consistent with [this](https://github.com/kubevirt/cluster-network-addons-operator/blob/547ad444cd072db8493ce93c37809937904e83f8/pkg/controller/statusmanager/status_manager.go#L97) Get command. since I am using this `config` object I thought I mught as welll changed also the `Get` , and make it like the other `Get`s in this module. but if you think this is not related  here I can turn it back.

RamLavi replied:
no I don't mind a little piggy-banking, just wanted to make sure I'm not missing something of essence.

RamLavi reviewed (commented):
added so,e comments and questions

alonSadan commented:
removed level from message, fixed constant to be uper-case and mvoed the emitert to it's own package and added New() function [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/111a868b7a3c74174617fe332f36f9d130341460..498a82b6208c698050b6a2905392b1a37224c1cf)

fixed typo in commit message. and separated the imports [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/547ad444cd072db8493ce93c37809937904e83f8..e15ba2db9ddd5a59011ab7997fb31045772efb03)

alonSadan commented:
added another commit to add events on change in the networkAddonsConfig spec

qinqon reviewed (changes requested):
Let's create some high level functions with event formating so we split responsabilities, added a comment to it.

RamLavi commented:
/lgtm

qinqon commented on pkg/eventemitter/eventemitter.go:
```diff
@@ -0,0 +1,91 @@
+
+func (ee eventEmitter) EmitProgressingForConfig(configName string) {
+	config := ee.getConfigForEmitter(configName)
+	if config != nil {
```
The `config != nil` is already check at `EmitEvent`  we can remove this from this function and it's friends.

alonSadan replied:
Tnx. 
Done.

qinqon reviewed (changes requested):
All good, just last nit.

alonSadan commented:
I made some changes: 
- pods controller: I added spec-Modified-emit to the pods-contoroler, to better notify on changes to spec
- config-controller:  moved the the Modified-emit to the begging of the reconcile loop. Same reason.
-   fixed @qinqon  nit:)

alonSadan commented:
We need to decide if this is were we want our emitters to be place.

From what I see in my laptop the emitters follow the behavior of the status changes.  ( if status `Progressing` is changed to `true` from `false` then a `Progressing` event will be created i.e)

two points I noticed that I think require discussion:
- spec-modifications-events also created when the Spec is firstly introduced.
- Progressing-event is not created when components are being deleted, only when created.  

I am fine with thees two points, but would love to here more opinions.

there could be other points to discuss but it requires more QA time.

alonSadan commented:
/retest

alonSadan commented:
added some info about the events in the readme

phoracek commented:
/retest

alonSadan commented:
/retest

alonSadan commented:
removed  CLUSTER_NETWORK_ADDONS_OPERATOR_NAME name from names.go. it is redundant.

modified some Wrapf() functions , which used `"...%s",err..`  since it duplicates the error message in the logs

qinqon commented:
/lgtm
If we need add/modify/delete events we can do at follow up PRs.

RamLavi commented:
/lgtm
/approve

phoracek commented:
/retest

phoracek commented:
/approve

phoracek commented on test/e2e/workflow/deployment_test.go:
```diff
@@ -151,25 +154,31 @@ var _ = Describe("NetworkAddonsConfig", func() {
 		It("should be able to remove all of them by removing the config", func() {
 			DeleteConfig()
 			CheckComponentsRemoval(components)
+			CheckModifiedEvent()
```
We don't need to test it in every single test case. In general, you want to have one test for each feature. When it is done like that, it is easier to identify which feature is broken.

Do we have a suiting tests for status changes? You should make sure that we don't test events multiple times in identical scenarios

Alternatively

alonSadan replied:
> We don't need to test it in every single test case. In general, you want to have one test for each feature. When it is done like that, it is easier to identify which feature is broken.
> 

I understand. So I think the recovery test covers all types of event creation. I'll keep only this now
> Do we have a suiting tests for status changes?

I think this is tested along with the integration tests. I don't think there is a  dedicated suit if that's what you have meant.

phoracek commented on test/check/events.go:
```diff
@@ -0,0 +1,211 @@
+package check
```
What is the source of this module?

alonSadan replied:
Sorry forgot to add it to the commit message. Added it now.
https://github.com/kubevirt/kubevirt/blob/master/tests/utils.go#L234

alonSadan commented:
/retest

qinqon commented:
/lgtm
